Mathlib/Analysis/Calculus/PartialDerivatives/Notation.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2025 Paul Lezeau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Eric Wieser
+-/
+import Mathlib.Analysis.Calculus.LineDeriv.Basic
+import Mathlib.LinearAlgebra.Basis.HasCanonicalBasis
+
+
+/-! # Notation for partial derivatives
+
+This file introduces some custom notation for partial derivatives. In order to make this
+as flexible as possible (e.g. we would like to have consistent notation that works for
+`ℝ × ℝ`, `Fin 2 → ℝ` and `EuclideanSpace ℝ (Fin 2)`), we work on abstract vector spaces
+equiped with a `HasCanonicalBasis` instance.
+
+We introduce two different notations
+- Whenever the index `i` is subscriptable, we can use `∂ᵢ[ℝ] f` to denote the `i-th` partial
+  derivative. This works for e.g `i : Fin n`.
+- In general, if the basis is indexed by a type `ι`, then we can use `∂[i; ℝ] f` for the
+  `i`-th partial derivative.
+
+-/
+
+universe u v w u' v' w'
+
+namespace HasCanonicalBasis
+/--
+The partial derivative of a function `F : V → 𝕜`.
+
+`partialDeriv 𝕜 F i tx` return the partial derivative of function `F : V → 𝕜`
+with respect to the `i`-th component of the canonical basis of `V` at point `tx : V`.
+
+This can be written using the following notations:
+- `(∂ᵢ[ℝ] f) tx` if the index `i` is subscriptable,
+- `(∂[i;ℝ] f) tx` in general.
+-/
+@[simp, nolint unusedArguments] -- we need `HasCanonicalBasis 𝕜 V ι f` to make the notation work
+noncomputable def partialDeriv (𝕜 : Type u) {V : Type v} {ι : Type w} {E : Type*}
+    {f : ι → V} [NontriviallyNormedField 𝕜] [AddCommGroup V]
+    [Module 𝕜 V] [HasCanonicalBasis 𝕜 V ι f] [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+    (F : V → E) (i : ι) (tx : V) : E :=
+  lineDeriv 𝕜 F tx (f i)
+
+section Notation
+
+open Mathlib Lean Meta Tactic Elab
+
+/-
+Define the partial derivative notations. Note that we need to use a macro
+rather than the `notation` command since we need to use the subscript parser
+`subscriptTerm`.
+-/
+
+@[inherit_doc partialDeriv]
+scoped syntax:100 "∂" noWs subscriptTerm noWs "[" term "]" term:max : term
+
+@[inherit_doc partialDeriv]
+scoped syntax:100 "∂[" term ";" term "]" term:max : term
+
+macro_rules
+  | `(∂$i:subscript[$𝕜] $f) => `(partialDeriv $𝕜 $f $i)
+  | `(∂[$i:term; $𝕜:term] $f) => `(partialDeriv $𝕜 $f $i)
+
```

Comment:
Another suggestion for the notation is to remove the `[𝕜]` argument

In SciLean, the derivative notation does not state the base field explicitly. It is expected that you usually work with one base field and state that with `set_default_scalar  𝕜` at the top of the file which just defines local notation `defaultScalar%` that expands to `𝕜` and the notation uses `defaultScalar%`
```
macro "set_default_scalar" r:term : command =>
  `(local macro_rules | `(defaultScalar%) => do pure (← `($r)).raw)
```

This is Kyle's solution discussed here https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/notation.20over.20field/with/419274309 

Adding something like `defaultScalar%` should be done as separate PR and then we can add an option to omit `𝕜` in the notation.